### PR TITLE
storage: Exclude eMMC "boot" non-partitions from multipath logic

### LINKED ISSUE #20250: Cannot control mmc device
### Explain what happens

hi, i have radxa x2l SBC with EMMC on board, but i cannot format, delete partitions or make another things using cockpit
image
when i click on this device i get this:
image
control buttons not available, please help me to debug and fix this problem, thanks, bet regards.
cockpit version:
image

### Version of Cockpit

287-1

### Where is the problem in Cockpit?

Storage

### Server operating system

Debian

### Server operating system version

12

### What browsers are you using?

Firefox

### System log

_No response_

### PR DESCRIPTION
**Pull request by mvollmer**
eMMC devices can have "boot" sub-devices which should be excluded from multipath considerations, just like partitions.

Fixes #20250

### CODE REVIEW

**Comment by mvollmer** [comment]
I have no idea how to test this... We might have to be content with not breaking anything that we can test, and getting feedback from the reporter of #20250.

**Comment by martinpitt** [comment]
Agreed - that's why we have packit COPRs, for easy user testing :grin:

File: pkg/storaged/client.js [hunk 1]
```diff
@@ -411,6 +396,41 @@ function btrfs_start_polling() {
...
+
+    if (client.blocks_part[block.path])
+        return false;
+
+    // Also, eMMCs have special partition-like sub-devices that point
+    // to the main Drive. We identify them by their name, just like
+    // UDisks2.
+
+    if (utils.decode_filename(block.Device).match(/^mmcblk[0-9]boot[0-9]$/))
+        return false;
```
Comment (cockpituous) [inline pkg/storaged/client.js:429]: This added line is not executed by any test.

**Comment by jelly** [comment]
I can reproduce this on my Surface:

It has an interesting bug with the empty card
Screenshot from 2024-04-23 15-28-25

With the patch:

No changes, let me re-check if I patched it correctly.

File: pkg/storaged/client.js [hunk 2]
```diff
@@ -411,6 +396,41 @@ function btrfs_start_polling() {
...
+    // as the object for the partition table. We have to ignore them.
+
+    if (client.blocks_part[block.path])
+        return false;
+
+    // Also, eMMCs have special partition-like sub-devices that point
+    // to the main Drive. We identify them by their name, just like
+    // UDisks2.
+
+    if (utils.decode_filename(block.Device).match(/^mmcblk[0-9]boot[0-9]$/))
```
Comment (jelly) [inline pkg/storaged/client.js:428]: This won't work, the devices have a `/dev/` prefix

Comment (jelly) [reply pkg/storaged/client.js:428]:
```
-    if (utils.decode_filename(block.Device).match(/^mmcblk[0-9]boot[0-9]$/))
+    if (utils.decode_filename(block.Device).match(/\/dev\/mmcblk[0-9]boot[0-9]$/))
```

This works.

Comment (jelly) [reply pkg/storaged/client.js:428]: Also our `add_disk` code does this:

```
 559   │         if not serial:
 560   │             serial = f"DISK{index}"
 561   │         dev = 'sd' + string.ascii_lowercase[index]
 562   │         extra = "<boot order='1'/>" if boot_disk else ""
 563   │         disk_desc = TEST_DISK_XML % {
 564   │             'file': image,
 565   │             'serial': serial,
 566   │             'unit': index,
 567   │             'dev': dev,
 568   │             'type': image_type,
 569   │             'extra': extra,
 570   │         }
```

So the device name seems user controlled, we could add a test for that? Or maybe via the `ram_disk` route?

Comment (mvollmer) [reply pkg/storaged/client.js:428]: > So the device name seems user controlled

Not really... I think the `dev = 'sd' + index` is just a good guess at what the device will be called in the guest. It doesn't control it.

Comment (mvollmer) [reply pkg/storaged/client.js:428]: > This won't work, the devices have a `/dev/` prefix

Ouch, that's why test, I guess. :-) Thanks!

**Review by jelly — APPROVED** [review]
Works!

**Pull request merged**